server/src/main/java/org/elasticsearch/common/util/concurrent/KeyedLock.java

```diff
@@ -63,20 +63,49 @@ public Releasable acquire(T key) {
         while (true) {
             KeyLock perNodeLock = map.get(key);
             if (perNodeLock == null) {
-                KeyLock newLock = new KeyLock(fair);
-                perNodeLock = map.putIfAbsent(key, newLock);
-                if (perNodeLock == null) {
-                    newLock.lock();
-                    return new ReleasableLock(key, newLock);
+                ReleasableLock newLock = tryCreateNewLock(key);
+                if (newLock != null) {
+                    return newLock;
+                }
+            } else {
+                assert perNodeLock != null;
+                int i = perNodeLock.count.get();
+                if (i > 0 && perNodeLock.count.compareAndSet(i, i + 1)) {
+                    perNodeLock.lock();
+                    return new ReleasableLock(key, perNodeLock);
                 }
             }
-            assert perNodeLock != null;
-            int i = perNodeLock.count.get();
-            if (i > 0 && perNodeLock.count.compareAndSet(i, i + 1)) {
-                perNodeLock.lock();
-                return new ReleasableLock(key, perNodeLock);
+        }
+    }
+
+    /**
+     * Tries to acquire the lock for the given key and returns it. If the lock can't be acquired null is returned.
+     */
+    public Releasable tryAcquire(T key) {
+        KeyLock perNodeLock = map.get(key);
+        if (perNodeLock == null) {
+            return tryCreateNewLock(key);
+        }
+        if (perNodeLock.tryLock()) { // ok we got it - make sure we increment it accordingly otherwise release it again
+            int i;
+            while ((i = perNodeLock.count.get()) > 0) {
```

Comment:
Is this necessary? Shouldn't we be able to return the `ReleasableLock` once we have acquired `perNodeLock`?

Reply:
I left a comment regarding this